docs/src/rfc/text/0017-aml-parser.md

```diff
@@ -0,0 +1,331 @@
+    parent_end: Option<usize>,
+}
+
+impl AmlSdtHandleInternal {
+    fn new(table_key: TableKey, offset: usize, size: usize) -> Self {
+        Self {
+            table_key,
+            offset,
+            size,
+            modified: false,
+            byte_encoding: AmlByteEncoding::default(),
+            parent_end: None,
+        }
+    }
+}
+
+pub type AmlHandle = AmlSdtHandleInternal;
+```
+
+The `size` of an `AmlSdtHandleInternal` refers to its full size, including any children (`TermList`).
+
+The `offset` refers to its offset with the AML stream of the table.
+Offset 0 is the start of the AML stream, and the highest offset is at the end of `table_length`.
+
+`modified` ensures the corresponding table (which can be retrieved through `table_key`) has an updated checksum
+if the contents are modified by `set_option`.
+
+Each handle stores the `parent_end` of its parent node, which is the parent's `size` + `offset` (useful for retrieving siblings).
+
+Each handle also stores its own byte encoding, specifying its own layout:
+
+```rust
+#[derive(PartialEq, Eq, Hash)]
+struct AmlByteEncoding {
+    opcode: AmlOpcode,
+    operands: Vec<AmlOperand>,
+    attributes: AmlOpAttributes,
+}
+
+bitflags! {
+    pub struct AmlOpAttributes: u32 {
+        /// If opcode has a pkg_length field.
+        const HAS_PKG_LENGTH  = 0x0000_0001;
+
+        /// If opcode has children.
+        const HAS_CHILD_OBJ   = 0x0000_0002;
+    }
+}
+
+/// Represents the possible opcodes. 
+pub enum AmlOpcode {
+    BaseOp(BaseOpcode),
+    ExtOp(ExtOpcode),
+}
+
+pub enum BaseOpcode {
+    ZeroOp,
+    AliasOp,
+    ...
+}
+
+pub enum ExtOpcode {
+    MutexOp,
+    EventOp,
+    ...
+}
+
+pub enum AmlOperand {
+    Opcode(AmlOpcode),
+    Name(AmlNameString), // Represents a NameString (AML path). Not to be confused with a string literal 
+    ...
+}
+```
+
+### AML Trait Interface
+
+The `AmlParser` service generally derives from the ACPI SDT protocol, and allows for traversal of the AML object tree.
+
+```rust
+pub(crate) trait AmlParser {  
+  // Opens a table's AML stream for parsing. The table should be a DSDT or SSDT. 
+  // The resulting handle is an opaque object on which further AML operations can be performed.
+  // It points to the first (root) node in the AML stream.
+  unsafe fn open_table(&self, table_key: TableKey) -> Result<AmlHandle, AmlError>;
+
+  // Closes a handle for modification and traversal.
+  // The handle will no longer be valid after it is closed.
+  // Update the corresponding table's checksum if the handle is `modified`. 
+  fn close_handle(&self, handle: AmlHandle) -> Result<(), AmlError>;
+
+  // Iterates over the options (operands) of an opened AML handle.
+  fn iter_options(&self, handle: AmlHandle) -> Result<Vec<AmlOperand>, AmlError>;
+
+  // Sets the option (operand) at a particular index to the given value.
+  fn set_option(&self, handle: AmlHandle, idx: usize, new_val: AmlOperand) -> Result<(), AmlError>;

```

Comment:
I know the "option" nomenclature comes from the SDT protocol but it seems out of place here. For example, what about something like `set_operand()`?